src/systems/gimbal_controller/GimbalControllerPlugin.cc

```diff
@@ -0,0 +1,490 @@
+{
+   if(!this->dataPtr->imuInitialized)
+   {
+      // Set unconditionally because we're only going to try this once.
+      this->dataPtr->imuInitialized = true;
+
+      auto imuEntity = _ecm.EntityByComponents(
+          components::Name(this->dataPtr->imuName),
+          components::Imu(),
+          components::ParentEntity(this->dataPtr->modelLink));
+
+      // Imu sensor topic
+      std::string topic = scopedName(imuEntity, _ecm) + "/imu";
+
+      if(topic.empty())
+      {
+         ignerr << "imu_sensor [" << this->dataPtr->imuName
+                << "] not found, abort ArduPilot plugin." << "\n";
+         return;
+      }
+      this->dataPtr->node.Subscribe(topic,
+         &GimbalControllerPluginPrivate::imuCb);
+   }
+
+   if(!this->dataPtr->pitchJoint || !this->dataPtr->rollJoint ||
+      !this->dataPtr->yawJoint)
+      return;
+
+   if(_info.simTime < this->dataPtr->lastControllerUpdateTime)
+   {
+      ignerr << "time reset event\n";
+      this->dataPtr->lastControllerUpdateTime = _info.simTime;
+      return;
+   }
+   else if(_info.simTime > this->dataPtr->lastControllerUpdateTime)
+   {
+      // Time delta
+      std::chrono::duration<double> dt =
+      (this->dataPtr->lastControllerUpdateTime - _info.simTime);
+
+      // pitch joint position
+      auto pitchJointPos = _ecm.Component<components::JointPosition>(
+         this->dataPtr->pitchJoint)->Data().at{0};
+
+      ignition::gazebo::components::JointForceCmd* pjfcComp = nullptr;
+      pjfcComp = _ecm.Component<components::JointForceCmd>(
+         this->dataPtr->pitchJoint);
+
+      if (pjfcComp == nullptr)
+      {
+         pjfcComp = _ecm.Component<components::JointForceCmd>(
+            _ecm.CreateComponent(this->dataPtr->pitchJoint,
+               components::JointForceCmd({0})));
+      }
+
+      // yaw joint position
+      auto yawJointPos = _ecm.Component<components::JointPosition>(
+         this->dataPtr->yawJoint)->Data().at{0};
+
+      ignition::gazebo::components::JointForceCmd* yjfcComp = nullptr;
+      yjfcComp = _ecm.Component<components::JointForceCmd>(
+         this->dataPtr->yawJoint);
+
+      if (yjfcComp == nullptr)
+      {
+         yjfcComp = _ecm.Component<components::JointForceCmd>(
+            _ecm.CreateComponent(this->dataPtr->yawJoint,
+               components::JointForceCmd({0})));
+      }
+
+      // roll joint position
+      auto rollJointPos = _ecm.Component<components::JointPosition>(
+         this->dataPtr->rollJoint)->Data().at{0};
+
+      ignition::gazebo::components::JointForceCmd* rjfcComp = nullptr;
+      rjfcComp = _ecm.Component<components::JointForceCmd>(
+         this->dataPtr->rollJoint);
+
+      if (rjfcComp == nullptr)
+      {
+         rjfcComp = _ecm.Component<components::JointForceCmd>(
+         _ecm.CreateComponent(this->dataPtr->rollJoint,
+            components::JointForceCmd({0})));
+      }
+
+      static int i = 1000;
+      if (++i > 100)
```

Comment:
why are you publishing 100 times the same value ?

Reply:
Referring to the plugin [file](https://github.com/addy1997/gazebo/blob/gazebo11/plugins/GimbalSmall2dPlugin.cc) in the gazebo classic, I think this portion is for publishing the **message** contents in form of string.